### DISCUSSION TITLE
Long time release connect to pool from context manager

### DISCUSSION THREAD

**Discussion by shurshilov** [sqlalchemy/sqlalchemy#10677] [Usage Questions] [answered]
Hi, i continue test your orm and my
![image](https://github.com/sqlalchemy/sqlalchemy/assets/11828278/63ec30ab-ceed-4192-abf5-8e4f6c812627)
And i found one strange, if i use conttext manager instead 28ms i get 78ms and only 50ms on aexit from context manage...
It look like you close conn instead just release it to pool, can explain why? because i just release conn and it work...ad not take many time...
And tell please isolation level by default autocommit as i understand?

**Comment by CaselIT** [comment]
hi,

> It look like you close conn instead just release it to pool,

not sure what you configured but by default sqlalchemy uses a queue pool so it does not close the connection.

>  And tell please isolation level by default autocommit as i understand?

Not sure where you read it, but the default isolation level is the default one obtained from database. You have to manually specify it if you want autocommit

**Comment by zzzeek** [comment, marked as answer]
outside of your block the session being closed and releasing the connection will be performing a rollback + connection pool checkin

> And tell please isolation level by default autocommit as i understand?

not at all, isolation level default comes from your driver and it is either "read committed" or "repeatable read", but never .autocommit, see https://peps.python.org/pep-0249/#commit  "Note that if the database supports an auto-commit feature, this must be initially off."

to set autocommit isolation levels with the ORM see https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#setting-transaction-isolation-levels-dbapi-autocommit

**Comment by shurshilov** [comment]
Yes, indeed, the autocomit was included in the test database. By the way, for reading, as I understand it, this is the fastest mode in mysql. 

**Can we not do a rollback and not check the connection pool, in the documentation of the drivers, (and in some prod projects) they do not do this and it seems to work stably?**

**Reply by zzzeek** [reply]
if you use autocommit mode as in the linked document, the connection.rollback() method will become a no-op and you will get all the perf gains associated with that.  

there is also the [reset-on-return](https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.Pool.params.reset_on_return) parameter that can be used to disable the rollback(), but you'd want to be in autocommit mode to use that anyway in modern use, so it won't make much difference.

**Reply by shurshilov** [reply]
Thank you, please answer the off-top. I saw an AsyncAdapter or something like that in your code. And maybe this is how you solved the problem that I have. 
Tested on mysql 8 and psotgres 12. 
I have two "update set ... where id=1" stmt. Autofixing mode is disabled. In two tables. These tables are not linked by FK. And I'm trying to execute a parallel transaction. 
I'm launching coroutines in gather. Each coroutine takes a connection from the pool, opens the cursor, due to the fact that autofixing is disabled, an implicit transaction opens, an update is performed and it does not close and the connection does not close. further, the second coroutine is also performed, but to the second table. Then, if there were no exceptions, I commit() both connections in the loop and return them to the pool. It works. And it works fast. But so far there are few requests, 1 in a few seconds. But as soon as I turn on load testing, the server freezes and lock occurs.
Do you know why?

**Reply by zzzeek** [reply]
if there's locks you'd want to look for deadlocks in your MySQL database, or there could be hangs in your program if you are using mutexes or something like that.   There's no way to know without a reproducing example

**Reply by shurshilov** [reply]
I wrote the code that reproduces this, if you click this endpoint in the browser it will work 100 out of 100. As soon as you load any tool for say 30 requests per second, the server will stop responding.

```py
import asyncio
import asyncpg
from fastapi import FastAPI
app = FastAPI()


class UnitOfWorkFlashORMPostgres:
    def __init__(self):
        self.session_factory = PosgresSessionWithTransaction

    async def __aenter__(self):
        self.session = self.session_factory()
        self.session.conn_list = []
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if exc_type is not None:
            for conn, tr in self.session.conn_list:
                await tr.rollback()
        else:
            for conn, tr in self.session.conn_list:
                await tr.commit()
                await conn.close()

class PosgresSessionWithTransaction:
    async def execute(
        self,
        stmt: str,
    ):
        try:
            conn: asyncpg.Connection = await asyncpg.connect(user='openpg', password='openpgpwd',
                                        database='newportal', host='127.0.0.1')
            tr = conn.transaction()
            await tr.start()
            await conn.execute(stmt)
            self.conn_list.append((conn, tr))

        except Exception as e:
            raise e


@app.on_event("startup") 
async def startup(): 
    create_table1="""CREATE TABLE IF NOT EXISTS message (id  serial  PRIMARY KEY,
    date       timestamp,
    subject         text,
    body   text,
    body_telegram        text,
    body_json         json,
    body_telegram_json json,
    chain_id integer,
    template_id integer
    )"""
    create_table2="""CREATE TABLE IF NOT EXISTS message_attributes (id  serial     PRIMARY KEY,
        date       timestamp,
        message_id integer,
        is_deleted boolean,
        is_draft boolean,
        shown_in_account boolean,
        send_email boolean
    )"""

    insert1 = "INSERT INTO message (date, subject) VALUES ('2023-11-23 10:00:00', 'Tampopo')"
    insert2 = "INSERT INTO message_attributes (date, message_id, is_deleted) VALUES ('2023-11-23 10:00:00', 1, true)"
    async with UnitOfWorkFlashORMPostgres() as uow:
        await uow.session.execute(create_table1)
        await uow.session.execute(create_table2)
        await uow.session.execute(insert1)
        await uow.session.execute(insert2)


@app.get("/")
async def index():
    async with UnitOfWorkFlashORMPostgres() as uow:
        st1="UPDATE message SET subject='test1', body='111' WHERE id = 1"
        st2="UPDATE message_attributes SET is_deleted=true WHERE message_id = 1"
        request_list = [uow.session.execute(st1),uow.session.execute(st2)]
        await asyncio.gather(*request_list)
```
start as 
python -m uvicorn test_lock:app --host 127.0.0.1 --port 81 --reload --http httptools --workers 1
